Readme.txt
% ===============================================================
The code in this package implements the Guided Image Denoising method for
image denoising as described in the following paper:

  Jun Xu, Lei Zhang, and David Zhang
  External Prior Guided Internal Prior Learning for Real-World Noisy Image Denoising.
  IEEE Transactions on Image Processing (TIP), 2018.

  Please cite the paper if you are using this code in your research.
  Please see the file License.txt for the license governing this code.

  Version:       1.0 (03/28/2018), see ChangeLog.txt
  Contact:       Jun Xu <csjunxu@comp.polyu.edu.hk/nankaimathxujun@gmail.com>
% ===============================================================

Note
------------
Please refer to https://github.com/csjunxu/GID_TIP2018 for the updates of the code.

Overview
------------
Training Code:
The code for learning external prior is provided in the folder "PG-GMM_TrainingCode", which relies
on the training images in the subfolder "Kodak24" (please refer to the "Data" section).

Testing Code:
The function "Demo_Guided_DND2017" demonstrates real-world image denoising with the Guided Image 
Denoising method introduced in the paper.

The function "Demo_Guided" demonstrates real-world image denoising with 
"ground truth" by the Guided Image Denoising method introduced in the paper.

The function "Demo_Guided_NoGT" demonstrates real-world image denoising 
without "ground truth" by the Guided Image Denoising method introduced in the paper.


Model
------------
The trained model "PGGMM_RGB_6x6_3_win15_nlsp10_delta0.001_cls33.mat" can be downloaded from
https://github.com/csjunxu/GID_TIP2018/PG-GMM_TrainingCode


Data
------------
Please download the data from corresponding addresses.
1. Kodak24: 24 high quality color images from Kodak PhotoCD dataset
                        This dataset can be found at http://r0k.us/graphics/kodak/
2. NCImages: real-world noisy images with no ''ground truth'' from "NoiseClinic"
                        This dataset can be found at http://demo.ipol.im/demo/125/

The "CCImages" directory include two parts:
3. CC15: 15 cropped real-world noisy images from CC [1]. 
                        This dataset can be found at  http://snam.ml/research/ccnoise
                        The smaller 15 cropped images can be found on in the directory 
                        ''Real_ccnoise_denoised_part'' of 
                        https://github.com/csjunxu/MCWNNM_ICCV2017
                                                The *real.png are noisy images;
                                                The *mean.png are "ground truth" images;
                                                The *ours.png are images denoised by CC.
4. CC60: 60 cropped (by us) real-world noisy images from CC [1]. 
                        "CC_60MeanImage" inlcudes the "ground truth" images;
                        "CC_60NoisyImage" inlcudes the noisy images;

5. DND_2017: 1000 cropped real-world noisy images from DND [2].
                         Please download the dataset from https://noise.visinf.tu-darmstadt.de/
                         and put the files in "DND_2017" directory accordingly.
6. PolyUImages: 100 cropped images from our new dataset.
                                                The *real.JPG are noisy images;
                                                The *mean.JPG are "ground truth" images;

[1] Seonghyeon Nam*, Youngbae Hwang*, Yasuyuki Matsushita, Seon Joo Kim. A Holistic Approach to 
      Cross-Channel Image Noise Modeling and its Application to Image Denoising. CVPR, 2016.
[2] Tobias Pl?tz and Stefan Roth. Benchmarking Denoising Algorithms with Real Photographs. CVPR, 2017.


Dependency
------------
This code is implemented purely in Matlab2014b and doesn't depends on any other toolbox.


Contact
------------
If you have questions, problems with the code, or find a bug, please let us know. Contact Jun Xu at 
csjunxu@comp.polyu.edu.hk or nankaimathxujun@gmail.com


ChangeLog.txt
28 Mar. 2018 - 1.0
- Initial release

Demo_Guided.m
clear;
%% 15 images of CC dataset
dataset = 'CC15';
GT_Original_image_dir = 'CCImages/CC15/';
GT_fpath = fullfile(GT_Original_image_dir, '*mean.png');
TT_Original_image_dir = 'CCImages/CC15/';
TT_fpath = fullfile(TT_Original_image_dir, '*real.png');

%% 60 images of CC dataset
% dataset = 'CC60';
% GT_Original_image_dir = 'CCImages/CC_60MeanImage/';
% GT_fpath = fullfile(GT_Original_image_dir, '*.png');
% TT_Original_image_dir = 'CCImages/CC_60NoisyImage/';
% TT_fpath = fullfile(TT_Original_image_dir, '*.png');

%% 100 images of our new dataset
% dataset = 'PolyU100';
% GT_Original_image_dir = 'PolyUImages/';
% GT_fpath = fullfile(GT_Original_image_dir, '*mean.JPG');
% TT_Original_image_dir = 'PolyUImages/';
% TT_fpath = fullfile(TT_Original_image_dir, '*real.JPG');

GT_im_dir  = dir(GT_fpath);
TT_im_dir  = dir(TT_fpath);
im_num = length(TT_im_dir);

method = 'GID';
write_MAT_dir = [dataset '_Results/'];
write_sRGB_dir = [write_MAT_dir method];
if ~isdir(write_sRGB_dir)
    mkdir(write_sRGB_dir)
end

load PG-GMM_TrainingCode/PGGMM_RGB_6x6_3_win15_nlsp10_delta0.001_cls33.mat;
% dictionary and regularization parameter
par.D= GMM.D;
par.S = GMM.S;
par.step = 3;       % the step of two neighbor patches
par.IteNum = 3;  % the iteration number
par.ps = ps;        % patch size
par.nlsp = nlsp;  % number of non-local patches
par.win = win;    % size of window around the patch
par.En = 54;

if strcmp(dataset, 'CC15')==1
    par.c1 = 0.001;
elseif strcmp(dataset, 'CC60')==1
    par.c1 = 0.001; % 0.0016 would be better
elseif strcmp(dataset, 'PolyU100')==1
    par.c1 = 0.001;
end
%         par.c2 = 0.005;
par.En = 54;
par.PSNR = [];
par.SSIM = [];
CCPSNR = [];
CCSSIM = [];
alltime = zeros(1,im_num);
for i = 1 : im_num
    IMin = im2double(imread(fullfile(TT_Original_image_dir, TT_im_dir(i).name) ));
    IM_GT = im2double(imread(fullfile(GT_Original_image_dir, GT_im_dir(i).name)));
    S = regexp(TT_im_dir(i).name, '\.', 'split');
    IMname = S{1};
    [h,w,ch] = size(IMin);
    fprintf('%s: \n',TT_im_dir(i).name);
    CCPSNR = [CCPSNR csnr( IMin*255,IM_GT*255, 0, 0 )];
    CCSSIM = [CCSSIM cal_ssim( IMin*255, IM_GT*255, 0, 0 )];
    fprintf('The initial PSNR = %2.4f, SSIM = %2.4f. \n', csnr( IMin*255,IM_GT*255, 0, 0 ), cal_ssim( IMin*255, IM_GT*255, 0, 0 ));
    par.I = IM_GT;
    par.nim =   IMin;
    par.imIndex = i;
    % Guided denoising
    t1=clock;
    [IMout,par]  =  Denoising_Guided_EI(par,model);
    t2=clock;
    etime(t2,t1)
    alltime(par.imIndex)  = etime(t2,t1);
    % calc ulate the PSNR and SSIM
    par.PSNR(par.IteNum,par.imIndex) =   csnr( IMout*255, IM_GT*255, 0, 0 );
    par.SSIM(par.IteNum,par.imIndex)      =  cal_ssim( IMout*255, IM_GT*255, 0, 0 );
    fprintf('The final : PSNR = %2.4f, SSIM = %2.4f\n', par.PSNR(par.IteNum,par.imIndex),par.SSIM(par.IteNum,par.imIndex));
    % output
    imwrite(IMout, [write_sRGB_dir '/' method '_' dataset '_' IMname '.png']);
end
PSNR = par.PSNR;
SSIM = par.SSIM;
mPSNR = mean(par.PSNR,2);
mSSIM = mean(par.SSIM,2);
mtime  = mean(alltime);
mCCPSNR = mean(CCPSNR);
mCCSSIM = mean(CCSSIM);
matname = sprintf([write_MAT_dir method '_' dataset '.mat']);
save(matname,'PSNR','mPSNR','SSIM','mSSIM','CCPSNR','mCCPSNR','CCSSIM','mCCSSIM');
